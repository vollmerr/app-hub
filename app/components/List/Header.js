import styled from 'styled-components';

import theme from 'utils/theme';

const Header = styled.div`
  flex: 1 ${theme.list.headerHeight};
  display: flex;
  padding: 5px 5px 15px 5px;
  min-height: ${theme.list.headerHeight};
`;

export default Header;
