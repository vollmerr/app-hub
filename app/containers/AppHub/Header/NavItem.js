import styled from 'styled-components';

import theme from 'utils/theme';


const focusStyle = (props) => `
  background: ${props.dark ? theme.themeDarker : theme.neutralPrimary};
`;


const NavItem = styled.div`
  display: flex;
  align-items: center;
  height: ${theme.hub.headerHeight}px;
  ${(props) => props.isLink &&
    `&:hover,
    &:active,
    &:focus {
      ${focusStyle(props)}
    }`
  }
  ${(props) => props.checked &&
    focusStyle(props)
  }
`;


export default NavItem;
