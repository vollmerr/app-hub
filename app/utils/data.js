/**
 * Formats an item object to the correct format
 * for rendering
 *
 * @param {array} item      - item to format
 * @param {string} name     - internal field name for lookup
 * @param {object} field    - field format info
 * @param {object} enums    - display names for enums
 *
 * @return {string}         - formatted item
 */
export function formatItem(item, name, field, enums = {}) {
  let content = item[name];
  // convert all data to array for mapping over
  if (!Array.isArray(content)) {
    content = [content];
  }
  // go through each item, mapping to the correct format
  content = content.map((x) => {
    // is enum mapping
    if (enums[name]) {
      return enums[name][x];
    }
    // is date
    if (field.data && field.data.type === 'DATE') {
      return isNaN(Date.parse(x)) ? '' : new Date(x).toISOString().substr(0, 10);
    }
    return x;
  });
  // join them back as string seperated by commas
  return content.join(', ');
}

/**
 * Formats an array of item objects to the correct format
 * for rendering
 *
 * @param {array} items     - items to format
 * @param {object} fields   - lookup table of fields
 * @param {object} enums    - display names for enums
 *
 * @return {array}          - formatted items as objects of strings
 */
export function formatItems(items, fields, enums) {
  const formattedItems = [];

  items.forEach((item, i) => {
    formattedItems[i] = {};
    Object.keys(item).forEach((k) => {
      formattedItems[i][k] = formatItem(item, k, fields[k], enums);
    });
  });

  return formattedItems;
}

/**
 * Maps a object to the format Office-UI-Fabric-React DetailsList`s expect
 *
 * @param {object} obj    - object to map
 * @param {array} include - keys to include, if empty includes all
 * @param {array} exclude - keys to exclude
 *
 * @return {object}       - mapped object to use in List
 */
export function mapToColumns(obj, include = [], exclude = []) {
  const keys = include.length ?
    include :
    Object.keys(obj);

  return keys
    .filter((key) => !exclude.includes(key))
    .map((key) => ({
      key: obj[key].name,
      name: obj[key].label,
      fieldName: obj[key].name,
      isResizable: true,
      ariaLabel: obj[key].ariaLabel,
      minWidth: obj[key].minWidth || 100,
      maxWidth: obj[key].maxWidth || 200,
      data: obj[key].data || {},
    }));
}
