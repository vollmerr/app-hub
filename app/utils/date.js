export const dateWithoutTime = (date) => {
  if (date && !isNaN(new Date(date))) {
    const d = new Date(date);
    d.setHours(0, 0, 0, 0);
    return d;
  }
  return undefined;
};
