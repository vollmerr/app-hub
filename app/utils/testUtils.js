import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

/**
 * Tests a styled-component for basic tests
 * such as rending correctly and adopting attributes
 * @param {Object} Component    - React component under test
 * @param {Object} extend       - React component the component under test extends /  inherits from
 */
export function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

/**
 * Tests a components for basic mapDispatchToProps information
 * @param {func} mapDispatchToProps   - components mapDispatchToProps function
 * @param {object} actions            - functions in mapDispatchToProps to test
 */
export function testMapDispatchToProps(mapDispatchToProps, actions) {
  const onAction = (action) => (
    `on${action.charAt(0).toUpperCase()}${action.slice(1)}`
  );

  let dispatch;
  let dispatchProps;
  beforeEach(() => {
    dispatch = jest.fn();
    dispatchProps = mapDispatchToProps(dispatch);
  });

  Object.keys(actions).forEach((action) => {
    describe(onAction(action), () => {
      it('should be injected', () => {
        expect(dispatchProps[onAction(action)]).toBeDefined();
      });

      it('should dispatch when called', () => {
        dispatchProps[onAction(action)]();
        expect(dispatch).toHaveBeenCalled();
      });
    });
  });
}
