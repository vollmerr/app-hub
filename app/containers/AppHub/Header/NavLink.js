import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { CommandButton } from 'office-ui-fabric-react/lib/Button';

import Link from 'components/Link';
import theme from 'utils/theme';

import NavItem from './NavItem';


export const ButtonStyled = styled(CommandButton) `
  color: ${theme.white};
  display: flex;
  justify-content: center;
  vertical-align: middle;
  border-width: 0;
  padding: 0 ${(props) => props.padding || '0'};
  min-width: ${theme.hub.headerHeight}px;
  height: ${theme.hub.headerHeight}px;
  box-sizing: border-box;
  text-align: center;
  line-height: normal;
  overflow: visible;
  transition: background-color .467s cubic-bezier(.1,.9,.2,1) 34ms;
  pointer-events: all;
  font-size: ${(props) => props.size || '24px'};
  &:hover,
  i.ms-Icon,
  &:hover i.ms-Icon,
  &.is-checked i.ms-Icon {
    color: ${theme.white};
  }
`;


export function Button({ text, children, onClick, panel, ...props }) { //eslint-disable-line
  const newOnClick = onClick ? () => onClick(panel) : null;

  return (
    <ButtonStyled
      onClick={newOnClick}
      {...props}
    >
      {text}
      {children}
    </ButtonStyled>
  );
}


class NavLink extends React.PureComponent {
  render() {
    const {
      iconProps,
      onClick,
      to,
      href,
      dark,
      title,
      checked,
      ...props
    } = this.props;

    const isLink = !!(onClick || to || href);
    const newIconProps = iconProps && { style: { fontSize: '20px' }, ...iconProps };

    const itemProps = {
      isLink,
      dark,
      checked,
    };

    const buttonProps = {
      dark,
      title,
      checked,
      onClick,
      ariaLabel: title,
      iconProps: newIconProps,
      ...props,
    };

    // handle routing
    if (to || href) {
      return (
        <NavItem {...itemProps}>
          <Link to={to} href={href} >
            <Button {...buttonProps} />
          </Link>
        </NavItem>
      );
    }

    // no links
    return (
      <NavItem {...itemProps}>
        <Button {...buttonProps} />
      </NavItem>
    );
  }
}


const { func, bool, string, node, shape } = PropTypes;

NavLink.propTypes = {
  panel: string,
  title: string,
  iconProps: shape({
    iconName: string,
  }),
  onClick: func,
  text: string,
  to: string,
  href: string,
  children: node,
  dark: bool,
  checked: bool,
  padding: string,
};


export default NavLink;
