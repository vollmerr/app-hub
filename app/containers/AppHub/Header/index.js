import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import toJS from '../../../hocs/toJS';
import theme from '../../../utils/theme';

import meta from '../meta';
import * as selectors from '../selectors';
import * as actions from '../actions';
import * as C from '../constants';

import NavLink from './NavLink';
import Logo from './Logo';
import UserInfo from './UserInfo';


const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  height: ${theme.hub.headerHeight}px;
  box-sizing: border-box;
  background: ${theme.neutralDark};
`;


const Section = styled.div`
  display: flex;
`;


const Line = styled.div`
  width: 1px;
  margin: auto;
  background: ${theme.neutralSecondary};
  height: ${(props) => props.partial ?
    `calc(${theme.hub.headerHeight}px - 15px)` :
    `calc(${theme.hub.headerHeight}px - 2px)`};
`;


class Header extends React.PureComponent {
  /**
   * Handles clicking on a panel from the header, determines if should open or close
   */
  handlePanelClick = (name) => {
    const { onChangePanelOpen, onChangePanelSelected, panel } = this.props;
    const selected = name || panel.selected; // errors if null, so set a default
    // click on same panel or overlay
    if (!name || (panel.selected === name && panel.isOpen)) {
      // close it
      onChangePanelOpen(false);
    } else {
      // open the panel (don't dispatch unessecary actions if already open)
      if (!panel.isOpen) {
        onChangePanelOpen(true);
      }
      onChangePanelSelected(selected);
    }
  }

  /**
   * Renders a nav link
   * @param {object} props    - props to pass through to nav link
   *
   * @return {JSX}            - NavLink to render
   */
  renderLink = (props) => {
    const { panel } = this.props;
    const linkProps = {
      onClick: this.handlePanelClick,
      ...props,
    };

    if (panel) {
      linkProps.checked = props.panel === panel.selected && panel.isOpen;
    }

    return (
      <NavLink {...linkProps} />
    );
  }

  render() {
    const {
      app,
      user,
      view,
    } = this.props;

    const { name: userName } = user;
    const { name: appName } = app;
    const { isMobile } = view;

    const appMenu = {
      isVisible: isMobile && appName,
      props: {
        iconProps: {
          iconName: 'appMenu',
        },
        title: `${appName} Navigation`,
        panel: C.APP_NAV_PANEL,
      },
    };

    const appHubTitle = {
      isVisible: !isMobile || !appName,
      props: {
        to: '/',
        text: meta.title,
        title: meta.title,
        iconProps: {
          iconName: 'Apphub',
          style: { fontSize: '24px', marginRight: '10px' },
        },
        padding: '15px',
      },
    };

    const appTitle = {
      isVisible: appName,
      props: {
        to: 'app.path', // TODO!
        text: appName,
        title: appName,
        padding: '15px',
      },
    };

    const userInfo = {
      isVisible: !isMobile && userName,
      props: {
        name: userName,
        initials: userName ? userName.split(/[,. ]/, 2).map((x) => x[0].toUpperCase()).join('') : '',
      },
    };

    const helpPanelProps = {
      iconProps: {
        iconName: 'helpMenu',
      },
      title: 'Help Panel',
      panel: C.HELP_PANEL,
    };

    const appHubPanelProps = {
      dark: true,
      iconProps: {
        iconName: 'appHubMenu',
      },
      title: 'App Hub Navigation',
      panel: C.APPS_PANEL,
    };

    return (
      <Wrapper>
        <Section>
          {appMenu.isVisible && this.renderLink(appMenu.props)}

          {!isMobile && <Logo />}
          {!isMobile && <Line partial />}

          {appHubTitle.isVisible && this.renderLink(appHubTitle.props)}
          {!name && <Line partial={!isMobile} />}

          {appTitle.isVisible && this.renderLink(appTitle.props)}
        </Section>

        <Section>
          {userInfo.isVisible && <UserInfo {...userInfo.props} />}
          <Line />

          {/* TODO: DEV/MOCK PANELS (+ Line) */}

          {/* TODO: ALERTS PANEL (+ Line) */}

          {this.renderLink(helpPanelProps)}
          <Line />

          {this.renderLink(appHubPanelProps)}
        </Section>
      </Wrapper>
    );
  }
}


const { func, object } = PropTypes;

Header.propTypes = {
  app: object.isRequired,
  user: object.isRequired,
  view: object.isRequired,
  panel: object.isRequired,
  onChangePanelOpen: func.isRequired,
  onChangePanelSelected: func.isRequired,
};


const mapStateToProps = createStructuredSelector({
  app: selectors.getApp,
  user: selectors.getUser,
  view: selectors.getView,
  panel: selectors.getViewPanel,
});

const mapDispatchToProps = (dispatch) => ({
  onChangePanelOpen: (open) => dispatch(actions.changePanelOpen(open)),
  onChangePanelSelected: (panel) => dispatch(actions.changePanelSelected(panel)),
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);


export default compose(
  withConnect,
  toJS,
)(Header);
