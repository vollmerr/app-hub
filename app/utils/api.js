import { delay } from 'redux-saga';
import { call, put } from 'redux-saga/effects';
import decode from 'jwt-decode';
import 'whatwg-fetch';

import { authUserDone } from '../containers/AppHub/actions';


export const DEFAULT_EXPIRE = 1000;
export const TOKEN = 'id_token';


/**
 * Determines if should fetch based off a cache time
 *
 * @param {Date} date             - date of cache
 * @param {Number} cacheSeconds   - number of seconds to cache
 *
 * @return {Bool}                 - if should fetch
 */
export function shouldFetch(date, cacheSeconds = 350) {
  return !date || (new Date(date).getTime() - (cacheSeconds * 1000)) > new Date().getTime();
}


/**
 * Sets a token for authenication in localStorage
 *
 * @param {String} token  - token to set
 */
export function setToken(token) {
  localStorage.setItem(TOKEN, token);
}


/**
 * Gets the authenication token from localStorage
 * @return {String}     - authenication token
 */
export function getToken() {
  return localStorage.getItem(TOKEN);
}


/**
 * Clears the authenication token from localStorage
 */
export function clearToken() {
  localStorage.removeItem(TOKEN);
}


/**
 * Helper function to valdate that token exists and is not expired
 *
 * @param {object} token  - jwt token
 *
 * @return {bool}         - valid ? true : false
 */
export function validToken(token) {
  try {
    const { exp } = decode(token);
    const now = new Date().getTime().toString().substring(0, 10);

    return Number(exp) > Number(now);
  } catch (e) {
    return false;
  }
}


/**
 * Helper function to put jwt token into global redux store
 *
 * @param {String} token    - jwt token
 *
 * @return {Number}         - exipration time
 */
export function* putToken(token) {
  try {
    const {
      sub: sam,
      exp,
      iat,
      ...rest
    } = decode(token);

    const expire = (exp - iat) * 1000; // convert from ms to s

    yield put(authUserDone({ sam, expire, ...rest }));
    return expire;
  } catch (error) {
    throw new Error(error);
  }
}


/**
 * Authenticates a user
 * Checks local storage for token first
 *
 * @param {Object} action   - redux action to dispatch (must have 'type')
 * @param {Number} maxTries - max tries to re attempt authenicating on fail
 */
export function* authenticate(action, maxTries = 3) {
  let expire = DEFAULT_EXPIRE; // try again in 1s by default
  let tries = 0;
  let error = new Error('Failed to authenticate');
  while (tries < maxTries) { // eslint-disable-line
    const token = getToken();
    // if valid local token use that, otherwise get new token
    if (validToken(token)) {
      // update user in global appHub state
      expire = yield call(putToken, token);
      // reset max tries
      tries = 0;
    } else {
      try {
        const options = {
          method: 'get',
          credentials: 'include',
        };
        const { id_token } = yield call(request, API.JWT, options); // eslint-disable-line
        // set into local storage for future authentication caching
        setToken(id_token);
        // update user in global appHub state
        expire = yield call(putToken, id_token);
        // reset max tries
        tries = 0;
      } catch (err) {
        // increment max tries
        tries += 1;
        error = err;
      }
      // max number of tires, set error
      if (tries === maxTries) {
        yield put(authUserDone(error));
      }
    }
    // authenicate again when token has expired
    yield call(delay, expire);
  }
}


/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
export function parseJSON(response) {
  if (response.status === 204 || response.status === 205) {
    return null;
  }
  return response.json();
}


/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
export function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error = new Error(response.statusText);
  error.response = response;
  throw error;
}


/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function request(url, options) {
  return fetch(url, options)
    .then(checkStatus)
    .then(parseJSON);
}


/**
 * Requests a URL with jwt token in header for authorization, returning a promise
 * Waits until jwt token is set in localStorage before making call
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function requestWithToken(url, options = {}) {
  let timer = null;

  const requestIfToken = () => { // eslint-disable-line
    // get users jwt token form storage else they dont have one
    const token = localStorage.getItem('id_token') || null;
    // valid token, call api
    if (validToken(token)) {
      clearTimeout(timer);
      // config settings to send api
      const newOptions = {
        method: options.method || 'get',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
        ...options,
      };
      // append optional body to request (POSTs)
      if (options.body) {
        newOptions.body = JSON.stringify(options.body);
      }

      return request(url, newOptions);
    }
    // wait again until token set
    timer = setTimeout(requestIfToken, 100);
  };

  return requestIfToken();
}
