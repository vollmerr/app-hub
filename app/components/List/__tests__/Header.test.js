import { testStyledComponent } from 'utils/testUtils';

import Header from '../Header';

testStyledComponent(Header);
