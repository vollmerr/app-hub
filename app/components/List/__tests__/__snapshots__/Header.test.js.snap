// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header__Header /> should render correctly 1`] = `
.c0 {
  -webkit-flex: 1 50px;
  -ms-flex: 1 50px;
  flex: 1 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 0;
  height: 50px;
}

<div
  className="c0"
/>
`;
